### PR TITLE
Improve renderer debuggability

### PR DESCRIPTION
**Pull request by kchibisov**
Make the renderer more debuggable by adding extra logging and using `GL_KHR_debug` when running with the debug log level.

### CODE REVIEW

**Comment by kchibisov** [comment]
@chrisduerr You can test with `cargo run -- -o debug.renderer=gles2_pure -vv` and applying:

```patch
diff --git a/alacritty/src/renderer/platform.rs b/alacritty/src/renderer/platform.rs
index cb123793..05ba35c8 100644
--- a/alacritty/src/renderer/platform.rs
+++ b/alacritty/src/renderer/platform.rs
@@ -102,10 +102,10 @@ pub fn create_gl_context(
 ) -> GlutinResult<NotCurrentContext> {
     let debug = log::max_level() >= LevelFilter::Debug;
     let mut profiles = [
-        ContextAttributesBuilder::new()
-            .with_debug(debug)
-            .with_context_api(ContextApi::OpenGl(Some(Version::new(3, 3))))
-            .build(raw_window_handle),
+        // ContextAttributesBuilder::new()
+        //     .with_debug(debug)
+        //     .with_context_api(ContextApi::OpenGl(Some(Version::new(3, 3))))
+        //     .build(raw_window_handle),
         // Try gles before OpenGL 2.1 as it tends to be more stable.
         ContextAttributesBuilder::new()
             .with_debug(debug)
diff --git a/alacritty/src/renderer/text/atlas.rs b/alacritty/src/renderer/text/atlas.rs
index 9d4cb401..94d713a2 100644
--- a/alacritty/src/renderer/text/atlas.rs
+++ b/alacritty/src/renderer/text/atlas.rs
@@ -160,18 +160,18 @@ impl Atlas {
                     multicolor = false;
                     // Gles context doesn't allow uploading RGB data into RGBA texture, so need
                     // explicit copy.
-                    if self.is_gles_context {
-                        let mut new_buffer = Vec::with_capacity(buffer.len() / 3 * 4);
-                        for rgb in buffer.chunks_exact(3) {
-                            new_buffer.push(rgb[0]);
-                            new_buffer.push(rgb[1]);
-                            new_buffer.push(rgb[2]);
-                            new_buffer.push(u8::MAX);
-                        }
-                        (gl::RGBA, Cow::Owned(new_buffer))
-                    } else {
+                    // if !self.is_gles_context {
+                    //     let mut new_buffer = Vec::with_capacity(buffer.len() / 3 * 4);
+                    //     for rgb in buffer.chunks_exact(3) {
+                    //         new_buffer.push(rgb[0]);
+                    //         new_buffer.push(rgb[1]);
+                    //         new_buffer.push(rgb[2]);
+                    //         new_buffer.push(u8::MAX);
+                    //     }
+                    //     (gl::RGBA, Cow::Owned(new_buffer))
+                    // } else {
                         (gl::RGB, Cow::Borrowed(buffer))
-                    }
+                    // }
                 },
                 BitmapBuffer::Rgba(buffer) => {
                     multicolor = true;

```

File: alacritty/src/renderer/mod.rs [hunk 1]
```diff
@@ -131,6 +135,16 @@ impl Renderer {
             (text_renderer, rect_renderer)
         };
 
+        // Enable debug logging for OpenGL as well.
+        if log::max_level() >= log::LevelFilter::Debug && GlExtensions::contains("GL_KHR_debug") {
```
Comment (chrisduerr) [inline alacritty/src/renderer/mod.rs:139]:
```suggestion
        if log::max_level() >= LevelFilter::Debug && GlExtensions::contains("GL_KHR_debug") {
```

File: alacritty/src/renderer/mod.rs [hunk 2]
```diff
@@ -131,6 +135,16 @@ impl Renderer {
...
+            debug!("Enabled debug logging for renderer");
```
Comment (chrisduerr) [inline alacritty/src/renderer/mod.rs:140]: Why is this necessary? To see if `GL_KHR_debug` is supported?

Also this should be "debug logging for OpenGL", not renderer.

File: alacritty/src/renderer/mod.rs [hunk 3]
```diff
@@ -131,6 +135,16 @@ impl Renderer {
...
+            unsafe {
+                gl::Enable(gl::DEBUG_OUTPUT);
+                gl::Enable(gl::DEBUG_OUTPUT_SYNCHRONOUS);
+                gl::DebugMessageCallback(Some(gl_debug_log), std::ptr::null_mut());
```
Comment (chrisduerr) [inline alacritty/src/renderer/mod.rs:144]:
```suggestion
                gl::DebugMessageCallback(Some(gl_debug_log), ptr::null_mut());
```

File: alacritty/src/renderer/platform.rs [hunk 4]
```diff
@@ -69,6 +71,23 @@ pub fn pick_gl_config(
...
+                    write!(msg, "\n  hardware_accelerated: {:?}", gl_config.hardware_accelerated());
+                let _ = write!(msg, "\n  config_api: {:?}", gl_config.api());
+                let _ = write!(
+                    msg,
+                    "\n  supports_transparency: {:?}",
+                    gl_config.supports_transparency()
+                );
+                let _ = write!(msg, "\n  srgb_capable: {:?}", gl_config.srgb_capable());
+                debug!("{}", msg);
+            }
```
Comment (chrisduerr) [inline alacritty/src/renderer/platform.rs:89]: This seems like the worst possible way to write this. Wouldn't it be easier to just use `format!`?

**Review by chrisduerr — CHANGES_REQUESTED** [review]

File: alacritty/src/renderer/mod.rs (hunk 2 above)
Comment (kchibisov) [reply alacritty/src/renderer/mod.rs:140]: Yeah, I want to check if extension is here.

File: alacritty/src/renderer/platform.rs (hunk 4 above)
Comment (kchibisov) [reply alacritty/src/renderer/platform.rs]: I think I've found a worse alternative.

File: alacritty/src/renderer/platform.rs [hunk 5]
```diff
@@ -69,6 +70,24 @@ pub fn pick_gl_config(
...
         if let Some(gl_config) = gl_config {
+            debug!(
+                r#"Picked GL Config:
+  buffer_type: {:?}
+  alpha_size: {}
+  num_samples: {}
+  hardware_accelerated: {:?}
+  supports_transparency: {:?}
+  config_api: {:?}
+  srgb_capable: {}"#,
```
Comment (chrisduerr) [inline alacritty/src/renderer/platform.rs:81]: I feel like this would be much better like this:

```suggestion
                "Picked GL Config:\n\
                  buffer_type: {:?}\n\
...
```

**Review by chrisduerr — CHANGES_REQUESTED** [review]

**Review by chrisduerr — APPROVED** [review]
